### ISSUE TITLE
Functions with wrong type argument called in new tank models?

### ISSUE THREAD

**Issue by casella** [DLR-SR/ThermofluidStream#195]
Many of the recently added models, e.g. `ThermofluidStream.Boundaries.Tests.TankExtendedTest2` fail in OMC with this error
```
[ThermofluidStream 1.1.0-main/Boundaries/Internal/partialTank.mo:126:3-127:39:writable]
Error: Type mismatch for positional argument 1 in 
ThermofluidStream.Boundaries.Tests.TankExtendedTest2.tank.Medium.Liquid.density(state=tank.medium.state). 
The argument has type:
  tank.Medium.ThermodynamicState
expected type:
  tank.Medium.Liquid.ThermodynamicState
```
This is the offending code:
https://github.com/DLR-SR/ThermofluidStream/blob/38553ffdb79f8d65566b06c46249a14605eca1e9/ThermofluidStream/Boundaries/Internal/PartialTank.mo#L126-L129

where

https://github.com/DLR-SR/ThermofluidStream/blob/38553ffdb79f8d65566b06c46249a14605eca1e9/ThermofluidStream/Boundaries/Internal/PartialTank.mo#L3-L5

and

https://github.com/DLR-SR/ThermofluidStream/blob/38553ffdb79f8d65566b06c46249a14605eca1e9/ThermofluidStream/Boundaries/Internal/PartialTank.mo#L58

I'm not sure how Dymola handles this code, but as far as I understand, OMC is right to complain: `tank.medium.state` is a `ThermodynamicState` record from an instance of `BaseProperties` from package `ThermofluidStream.Media.myMedia.GasAndIncompressible.JP8DryAir`, while the function `density()` is called once from package `Medium.Liquid`, i.e. `ThermofluidStream.Media.myMedia.Incompressible.Examples.JP8` and once from package `Medium.Gas`, i.e. `ThermofluidStream.Media.myMedia.Air.DryAirNasa`. 

This cannot work: the `density()` function and the `BaseProperties` model instance should both be taken from the same package to be consistent, otherwise how could they compute the right density?

@perost, can you please double-check and comment?

As I understand, you should probably define `PartialGasAndIncompressible.BaseProperties` to contain two state records, e.g.
```modelica
Gas.ThermodynamicState state_gas;
Liquid.ThermodynamicState state_liquid;
```
then, you could have something like
```modelica
Modelica.Units.SI.Density liquidDensity=Medium.Liquid.density(medium.state_liquid) 
  "density of the liquid in the tank"; 
 Modelica.Units.SI.Density gasDensity=Medium.Gas.density(medium.state_gas) 
  "density of the gas in the tank"; 
```
of course PartialGasAndIncompressible.BaseProperties can also contain 
```modelica
ThermodynamicState state;
```
but this state should contain the information about the overall properties of the gas-liquid mixture.

**Comment by perost** [comment]
> @perost, can you please double-check and comment?

Yes, `Medium.Liquid.density` and `Medium.Gas.density` both expect a `ThermodynamicState` record that looks like:
```Modelica
record ThermodynamicState
  AbsolutePressure p;
  Temperature T;
end ThermodynamicState;
```
But `medium.state` is an instance of `JP8DryAir.ThermodynamicState` which looks like:
```Modelica
record ThermodynamicState
  AbsolutePressure p;
  Temperature T;
  MassFraction X[ns];
end ThermodynamicState;
```
I.e. it has an extra field `X`. The specification says that function arguments "must agree with the type of the corresponding parameter", which could mean pretty much anything. But we interpret it to mean that they must be type compatible as defined in 6.7, which states that the records must have the same named elements.

**Comment by IngelaLind** [comment]
Thank you for pointing this out, in Dymola there isn´t even a warning. I will look into it, but since it is outside my comfort zone, it might take a while to find a good solution. I will take your analysis into account.

From: Per Östlund ***@***.***>
Sent: den 19 juni 2024 23:34
To: DLR-SR/ThermofluidStream ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [DLR-SR/ThermofluidStream] Functions with wrong type argument called in new tank models? (Issue #195)


@perost<https://smex-ctp.trendmicro.com:443/wis/clicktime/v1/query?url=https%3a%2f%2fgithub.com%2fperost&umid=7863382c-fe87-44f5-a09c-f4d2131e78d4&auth=bf46f176bc8ebabeebda796db2a60a8ae677f9c4-bce46fe30b3bc165d7395c8c995aa20420d2155a>, can you please double-check and comment?

Yes, Medium.Liquid.density and Medium.Gas.density both expect a ThermodynamicState record that looks like:

record ThermodynamicState

  AbsolutePressure p;

  Temperature T;

end ThermodynamicState;

But medium.state is an instance of JP8DryAir.ThermodynamicState which looks like:

record ThermodynamicState

  AbsolutePressure p;

  Temperature T;

  MassFraction X[ns];

end ThermodynamicState;

I.e. it has an extra field X. The specification says that function arguments "must agree with the type of the corresponding parameter", which could mean pretty much anything. But we interpret it to mean that they must be type compatible as defined in 6.7, which states that the records must have the same named elements.

**Comment by IngelaLind** [comment]
I think the function call should be 
Modelica.Units.SI.Density liquidDensity=Medium.Liquid.density(Medium.Liquid.setState_pTX(medium.state.p,medium.state.T,{1}))
    "density of the liquid in the tank";
  Modelica.Units.SI.Density gasDensity=Medium.Gas.density(Medium.Gas.setState_pTX(medium.state.p,medium.state.T,{1}))
    "density of the gas in the tank";
This simulates in Dymola. But that did the previous version as well, without any warnings at all, even with pedantic check. Now lets see if I can find out how to suggest the change.

**Comment by casella** [comment]
Sounds good, it should also work in OMC. BTW, you should be able to skip the `{1}` input, as the `setState_pTX` function has a default input value `reference_X` for it.
